File: pages/LoginPage.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './BasePage';

/**
 * LoginPage — Page Object for /login
 * Covers TC-001 (valid login) and TC-002 (invalid login).
 */
export class LoginPage extends BasePage {
  readonly usernameInput: Locator;
  readonly passwordInput: Locator;
  readonly loginButton: Locator;
  readonly flashMessage: Locator;

  constructor(page: Page) {
    super(page);
    this.usernameInput = page.getByRole('textbox', { name: 'Username' });
    this.passwordInput = page.getByRole('textbox', { name: 'Password' });
    this.loginButton   = page.getByRole('button', { name: /Login/i });
    this.flashMessage  = page.locator('#flash');
  }

  async goto(): Promise<void> {
    await this.navigate('/login');
  }

  async login(username: string, password: string): Promise<void> {
    await this.usernameInput.fill(username);
    await this.passwordInput.fill(password);
    await this.loginButton.click();
  }

  async getFlashMessage(): Promise<string> {
    await this.flashMessage.waitFor({ state: 'visible' });
    return (await this.flashMessage.innerText()).trim();
  }
}

File: selenium-java-pom/src/main/java/com/ailearning/seleniumpom/pages/LoginPage.java
package com.ailearning.seleniumpom.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class LoginPage extends BasePage {
    private static final By USERNAME_INPUT = By.id("username");
    private static final By PASSWORD_INPUT = By.id("password");
    private static final By LOGIN_BUTTON = By.cssSelector("button[type='submit']");
    private static final By FLASH_MESSAGE = By.id("flash");

    public LoginPage(WebDriver driver) {
        super(driver);
    }

    public void gotoPage() {
        navigate("/login");
    }

    public void login(String username, String password) {
        waitForVisible(USERNAME_INPUT).sendKeys(username);
        waitForVisible(PASSWORD_INPUT).sendKeys(password);
        waitForVisible(LOGIN_BUTTON).click();
    }

    public String getFlashMessage() {
        return getNormalizedText(FLASH_MESSAGE);
    }
}
